rust/src/adt.rs

```diff
@@ -0,0 +1,846 @@
+// SPDX-License-Identifier: MIT
+use core::ffi::*;
+use core::mem::size_of;
+
+use crate::c_size_t;
+
+#[derive(Debug)]
+pub enum AdtError {
+    NotFound = -1,
+    BadOffset = -4,
+    BadPath = -5,
+    BadNCells = -14,
+    BadValue = -15,
+    BadLength = -20,
+}
+
+const ADT_ALIGN: i32 = 4;
+
+#[repr(C)]
+pub struct ADT<'a> {
+    size: usize,
+    data: &'a [u8],
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct ADTNode {
+    property_count: u32,
+    child_count: u32,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct ADTProperty {
+    name: [c_char; 32],
+    size: u32,
+    value: [u8],
+}
+
+#[repr(C, packed(1))]
+pub struct ADTSegmentRanges {
+    phys: u64,
+    iova: u64,
+    remap: u64,
+    size: u32,
+    unk: u32,
+}
+
+impl<'a> ADT<'a> {
+    /// Check that our byte offset into ADT memory is properly aligned
+    fn check_offset(offset: i32) -> Result<(), AdtError> {
+        if offset < 0 || (offset % ADT_ALIGN != 0) {
+            Err(AdtError::BadOffset)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Determine if an ADT node's name is equal to some string bounded by a length.
+    /// This is required to match the semantics expected when doing a full ADT path
+    /// trace.
+    fn node_names_equal(a: &str, b: &str, len: usize) -> bool {
+        if a == &b[..len] {
+            return true;
+        }
+
+        if b[..len].contains('@') && a.as_bytes().get(len) == Some(&b'@') {
+            return true;
+        }
+
+        false
+    }
+
+    /// Trace a given ADT path and return the offset of the final node.
+    pub fn path_offset_trace(path: &str, mut offsets: Option<&mut [i32]>) -> Result<i32, AdtError> {
```

Comment:
Here and everywhere else where you use node offsets - unsound api from Rust's POV as the offsets are invalidated when you change the ADT, but i am not sure how to properly model it for now and whether it is worth it.